### PR TITLE
Compute and use optimal batchsize

### PR DESCRIPTION
**Pull request by Leengit**
Closes Issue #22.

Whether tensorflow or torch, a supplied `args.batchSize` of -1 (or any other value less than 1) will be a request to optimize the batch size.  The result will be the largest power of two that doesn't cause an exception.  (Edit: we also introduce a timing metric.)

<strike>Note that this pull request is hardcoded to always optimize the batch size, and thus will optimize the batch size even though the code in the `wsi-superpixel-guided-labeling` repository is not yet updated to supply a value of -1 for `batchSize`.  This will enable testing of this pull request in isolation.</stike>  After testing, a subsequent commit (or separate pull request) can undo the hardcoded value of -1 for `batchSize`.

### CODE REVIEW

File: superpixel_classification/SuperpixelClassification/SuperpixelClassificationBase.py [hunk 1]
```diff
@@ -795,12 +795,14 @@ def main(self, args):
                 gc, args.images, args.annotationName, args.features, args.patchSize,
                 args.numWorkers, prog)
 
+            batchSize = args.batchSize
+            batchSize = -1  # Always optimize the batch size
```
Comment (manthey) [inline superpixel_classification/SuperpixelClassification/SuperpixelClassificationBase.py:799]: Is there a reason this is here, or is it leftover from testing?  And, I think we should change the default value in the xml.

Comment (Leengit) [reply superpixel_classification/SuperpixelClassification/SuperpixelClassificationBase.py:799]: I put it there so that it would be easy for reviewers to test.  I'll pull it out before we merge this.

Comment (Leengit) [reply superpixel_classification/SuperpixelClassification/SuperpixelClassificationBase.py:799]: > And, I think we should change the default value in the xml.

Will do.

**Comment by manthey** [comment]
In trying this out on a particular test set, it works but slows things down, seemingly where we are detecting the maximum batch size.  Specifically, on my test set using the Tensorflow class, it finds a maximum batch size of 8192.  When run with -1 to detect the batch size, it takes 55 to 56 seconds.  When run with 8192 explicitly, it takes 34 seconds.  Further, if I manually try different batch sizes, I find the fastest power of two value ends up being 256 which takes 22 seconds.

Obviously, we want to utilize the GPU effectively, but the goal is maximum speed, so I wonder if we measured the time to run the various batch sizes and stop as soon as the time per sample starts to increase?  A quick hack shows that this would have picked 128.  Timing is inherently noisy, so I don't know how reproducible this will be (that is, if the system is doing something else, we might stop early because of other load on the system).

My quick hack is measuring the time around the try/except block and dividing it by the batch size.  If that is higher than a previous run, return just as we would have for a memory error.

File: superpixel_classification/SuperpixelClassification/SuperpixelClassificationTensorflow.py [hunk 2]
```diff
@@ -87,6 +96,19 @@ def predictLabelsForItemDetails(self, batchSize, ds, item, model, prog):
         catWeights = tf.nn.softmax(predictions)
         return catWeights, predictions
 
+    def findOptimalBatchSize(self, model, ds) -> int:
+        # Find an optimal batch_size
+        maximum_batchSize: int = 65536  # power of two
+        batchSize: int = 1
+        while batchSize < maximum_batchSize:
```
Comment (manthey) [inline superpixel_classification/SuperpixelClassification/SuperpixelClassificationTensorflow.py:103]: When training the model, we might not have this many samples, so we should also stop when the batchsize is greater that the size of ds.

**Comment by manthey** [comment]
Since finding the optimal batch size takes time, we should only do it once for training and once for predictions.  Currently, this does it for each item that is being predicted.

File: superpixel_classification/SuperpixelClassification/SuperpixelClassificationTensorflow.py [hunk 3]
```diff
@@ -87,6 +97,29 @@ def predictLabelsForItemDetails(self, batchSize, ds, item, model, prog):
...
+                model.predict(small_ds, batch_size=batchSize)
+                elapsed_time = time.time() - start_time
+                if elapsed_time > 2 * previous_time + add_seconds:
+                    return batchSize // 2
+                previous_time = elapsed_time
+            except tf.errors.OpError:
+                return batchSize // 2
+            batchSize *= 2
+        # Undo the last doubling; it was spurious
+        return batchSize // 2
```
Comment (manthey) [inline superpixel_classification/SuperpixelClassification/SuperpixelClassificationTensorflow.py:121]: For the case where len(ds) is not a power of two, this will end up returning less than len(ds).  You had been doing the doubling at the start of the while block which would avoid that issue.

File: superpixel_classification/SuperpixelClassification/SuperpixelClassificationTensorflow.py [hunk 4]
```diff
@@ -87,6 +97,29 @@ def predictLabelsForItemDetails(self, batchSize, ds, item, model, prog):
         catWeights = tf.nn.softmax(predictions)
         return catWeights, predictions
 
+    def findOptimalBatchSize(self, model, ds) -> int:
+        # Find an optimal batch_size
+        maximum_batchSize: int = len(ds)
+        batchSize: int = 2
+        # We are using a value greater than 0.0 for add_seconds so that small imprecise
+        # timings for small batch sizes don't accidentally trip the time check.
+        add_seconds: float = 0.5
```
Comment (manthey) [inline superpixel_classification/SuperpixelClassification/SuperpixelClassificationTensorflow.py:106]: This is too large on my system -- it never invokes the time restriction.  I'd use a much smaller value (maybe 0.05)?

File: superpixel_classification/SuperpixelClassification/SuperpixelClassificationBase.py (hunk 1 above)
Comment (Leengit) [reply superpixel_classification/SuperpixelClassification/SuperpixelClassificationBase.py]: I've removed the setting of batchsize to -1.  It's all up to the caller now, as it should be, though the default is now -1 in the `xml` file.

File: superpixel_classification/SuperpixelClassification/SuperpixelClassificationTensorflow.py (hunk 2 above)
Comment (Leengit) [reply superpixel_classification/SuperpixelClassification/SuperpixelClassificationTensorflow.py]: Good, thank you.

File: superpixel_classification/SuperpixelClassification/SuperpixelClassificationTensorflow.py (hunk 3 above)
Comment (Leengit) [reply superpixel_classification/SuperpixelClassification/SuperpixelClassificationTensorflow.py:121]: Thank you, fixing that.

File: superpixel_classification/SuperpixelClassification/SuperpixelClassificationTensorflow.py (hunk 4 above)
Comment (Leengit) [reply superpixel_classification/SuperpixelClassification/SuperpixelClassificationTensorflow.py:106]: Thank you for sleuthing that out, though I'm puzzled... it works well for my example.  I think that you are saying that processing one batch for a given batch size never takes more than 0.5 second longer than twice the time for single batch that is half that size.  I think that means either that bigger is always better or, if not, that bigger is only barely worse; which isn't what you are seeing so I am confused.

What is your dataset size, is this for training or prediction, what run times are you seeing?

Comment (manthey) [reply superpixel_classification/SuperpixelClassification/SuperpixelClassificationTensorflow.py:106]: I was testing on a two image set, one with ~11,000 superpixels and the other with ~18,000 superpixels.  

Looking at the time to run the test batches for predication, I see (in seconds), on the first image, I see:
batchsize 2 0.23432278633117676 
batchsize 4 0.16736316680908203 
batchsize 8 0.0986783504486084 
batchsize 16 0.11266183853149414
batchsize 32 0.11878609657287598 
batchsize 64 0.11466693878173828
batchsize 128 0.13788604736328125
batchsize 256 0.5580830574035645 
I'd have to find a previous debug or re-add the debug statement to look beyond that.  But, here we want to stop at 128 but 0.13 * 2 + 0.5 is not less than 0.55.

This is on a system with a decent NVidia GPU.

Comment (Leengit) [reply superpixel_classification/SuperpixelClassification/SuperpixelClassificationTensorflow.py]: Okay, I've changed `add_seconds` to `0.05`.  Let's see how that works.

**Comment by Leengit** [comment]
> My quick hack is measuring the time around the try/except block and dividing it by the batch size. If that is higher than a previous run, return just as we would have for a memory error.

For posterity, nothing that I've pushed code that implements this, with `add_seconds` as we discuss elsewhere.

**Comment by Leengit** [comment]
> Since finding the optimal batch size takes time, we should only do it once for training and once for predictions. Currently, this does it for each item that is being predicted.

For posterity, noting that I pushed code to implement this.

**Review by manthey — APPROVED** [review]
